### PR TITLE
feat: [M13] Swift: an fp16/bf16 parity contract (a different, looser tolerance than fp32 1e-4)

### LINKED ISSUE #266: [M13] Swift: an fp16/bf16 parity contract (a different, looser tolerance than fp32 1e-4)
Follow-up from #166 (PR #262). The Swift port carries the full compute-dtype plumbing (`cast`/`f32`/`linear` route through `MambaConfig.cd`, fp32 stays the original op verbatim), but **only fp32 is gated** — every checked-in fixture is fp32, and `config/poc.yaml` is fp16-overridden-to-fp32 for the harness.

`docs/design/03-conformance.md` explains why fp32 1e-4 cannot simply be reused: bf16's epsilon is too coarse for the comparison to mean anything. So this needs a genuinely different contract, not a relaxed constant — and it is needed **before #167 ships a fast decode path at poc's native precision**, since that path is exactly the one nothing currently gates.

Scope: pick the contract (a looser abs/rel pair? top-1 agreement + KL, like #168's quant gate?), add fp16/bf16 fixtures via `scripts/export_parity_fixture.py --precision`, and extend `monica-parity` to apply the right contract per fixture. Keep the fp32 gates running unchanged against the fp model.

### PR DESCRIPTION
**Pull request by travisgalloway**
Closes #266

## Summary

Adds a genuinely separate, measured precision->tolerance contract for the Swift engine's low-precision fixtures, on top of #168's per-fixture `rtol`/`atol` hook (#168 built the minimal hook only; this issue builds the actual precision->tolerance table and the low-precision fixture family on top of it).

- `src/conformance/tolerances.py` (new, portable): the derived elementwise band (`atol=64u`, `rtol=8u`, `u` = unit roundoff), a load-bearing mean-KL tier for low precision, and dtype-scaled margin guards for the two exact comparisons (`greedy_ids`, MoE `load.{i}`). Full derivation + measured tables in the module docstring.
- `scripts/export_parity_fixture.py`: fixes the blocker found while implementing — `check_forward_step_parity`/`check_prefill_decode_parity` were called with no tolerance override, so both always used fp32 defaults and no low-precision fixture could ever be written. Also adds anti-vacuity refusals (a fixture whose own noise floor isn't comfortably inside its gate never reaches disk), a `--allow-moe-load-omit` declared-fallback for the route-margin guard, an `np.isfinite` overflow check, and a `_np_f32` helper working around a local MLX 0.32.0 buffer-protocol bug that blocked converting intermediate `bfloat16` arrays to numpy.
- `swift/engine/Sources/monica-parity/main.swift`: `precisionBands` table (mirrors `tolerances.py`), `FixtureMeta.precision` (missing/unrecognised is a hard failure, never a default), the override guard (only a `quant_bits` fixture may carry `rtol`/`atol`, and only ever loosens), the low-precision KL/top-1 distributional tier, and the P6 packing gate now reads the dtype band instead of hardcoded fp32 constants.
- Four new fixtures: `toy-fp16`, `toy-bf16`, `toy-moe-fp16` (seed 15 + `--moe-bias` — no seed at the default ranking cleared the fp16 route-margin guard on this toy model), `toy-hybrid-fp16`.
- `tests/test_lowp_parity_band.py` (new): T1 calibration reproduction, T2 anti-vacuity, T3 falsification (the KL tier actually rejects a defect), T4 the fp32 gate still fails when perturbed, T5 the structural fp32-non-regression guarantee (portable, no MLX — runs on the Linux CI job).
- `tests/test_parity_fixture_export.py`: extends the staleness check to `toy-fp16`.
- `tests/test_import_guard.py`: adds `src.conformance.tolerances` to `PORTABLE_MODULES`.
- `docs/design/03-conformance.md` / `swift/engine/Fixtures/README.md`: documents the low-precision contract as an *additional*, weaker gate — never a replacement for the fp32 one.

**Finding during implementation:** the plan's KL threshold (`0.1*u^2`, calibrated on `toy.yaml` alone) undershot the real noise floor by 40-60x for `toy-moe.yaml`/`toy-hybrid.yaml` — MoE routing and attention introduce coherent (non-cancelling) forward-vs-step divergence at low precision, not just rounding noise. Recalibrated on the worst of the three toy configs (the same worst-case-across-configs methodology the elementwise band already used), landing at `9e-7` (fp16) / `6e-5` (bf16). Detection floor is correspondingly weaker than the plan's original estimate (~2.8%/~3% vs. ~2.3%/~11%) but still ~12x/~110x more sensitive than the elementwise tier — still clearly load-bearing. Full derivation in `tolerances.py`'s docstring.

## Verification

- `.venv/bin/python -m pytest -q -rs`: **1595 passed, 11 skipped** (pre-existing/unrelated skips: no CUDA device, no `datatrove`, `torch.compile` unusable on this host).
- Fixture export: all four new fixtures exported and re-export-reproducible (`cmp`'d byte-identical across two runs each, fp16 and bf16 — no #298 buffer-cache-reuse interaction observed).
- `cd swift && swift build` — OK. `swift run monica-selfcheck` — OK. `swift package show-dependencies --format flatlist` — empty (confirms `swift/` stays dependency-free). `git diff main -- swift/Package.swift` — empty (no mlx-swift edge added).
- `cd swift/engine && swift build` — OK.
- `swift run monica-parity` — fails at the pre-existing `MLX error: Failed to load the default metallib` (Command Line Tools only, no Xcode on this host) *before* reaching any of my code — same failure mode as on the pre-#266 fixtures, so this is not a regression. The authoritative gate is CI's `swift model parity (macOS, mlx-swift)` job.
- **fp32 strict-gate non-regression, literal before/after** (`layers.0.out_proj.weight` on `toy.yaml`, in-memory only, no checked-in fixture touched):
  - BEFORE (no perturbation): `allclose(rtol=1e-4, atol=1e-5) = True`
  - AFTER (1e-2 relative weight defect): `max|Δlogit| = 1.234e-03`, `allclose(rtol=1e-4, atol=1e-5) = False`
  - No band was widened to make this (or any other) run pass.

## CI-only

- `swift model parity (macOS, mlx-swift)` — the authoritative Swift gate for all 12 fixtures (8 pre-existing + 4 new).
- `swift selfcheck (Linux, swift:latest)` and the Linux side of `swift package show-dependencies`.
- The portable Linux CI job additionally covers `tests/test_lowp_parity_band.py`'s T5 with no MLX installed.

## Testing
- [x] Tests added/updated
- [x] All tests passing (local suite; Swift gate is CI-only per above)
- [x] Manual testing completed (fixture export, reproducibility, fp32 regression before/after)

### CODE REVIEW

File: scripts/export_parity_fixture.py [hunk 1]
```diff
@@ -200,24 +237,71 @@ def build_fixture(config_path: str, out_dir: str, *, batch: int, seq: int,
...
+                f"rtol={rtol_fp}, atol={atol_fp}, precision={cfg.precision})")
 
         # Same rule for #169's state handoff: a fixture whose own Python prefill state
         # disagrees with its stepped state (or whose prefill logits disagree with forward, or
         # whose prefill-then-decode disagrees with pure step-by-step) can never reach disk —
         # it would make the Swift gate compare against an internally-inconsistent oracle.
         prefill_verdict = check_prefill_decode_parity(
-            ref_model, tokens, to_numpy=np.array, n_decode=min(4, seq - 1))
+            ref_model, tokens, to_numpy=np.array, n_decode=min(4, seq - 1),
+            rtol=rtol_fp, atol=atol_fp)
```
[the remark this answers is not in the document]
Comment (travisgalloway) [reply scripts/export_parity_fixture.py:265]: ⏭️ Reviewed - Not a bug: state leaves are fp32 regardless of `precision`, by explicit design, not built in compute dtype. `_conv_window` (mlx_backend.py) casts to fp32 ("Emitted in fp32 so it is dtype-identical to a stepped window"); the SSD carry-out state in `SelectiveSSM.recurrence`/`parallel` stays fp32 by construction ("scan + state stay fp32", "fp32 carry-out"); attention's k/v cache is likewise cast fp32 in `_qkv`'s `heads()`. Reproduced directly: built a toy model at `precision=bf16` and ran `check_prefill_decode_parity(model, tokens, to_numpy=np.array, ...)` — it succeeds with no buffer-protocol error, confirming the state dtype claim doesn't hold. `_np_f32` exists for a different, real problem (bf16 INTERMEDIATE activations like hidden states / mixing matrices, not prefill/step State). No change needed here.
